File: tsk_lib/src/list.rs
// Copyright 2018 Mathew Robinson <chasinglogic@gmail.com>. All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use chrono::prelude::*;

use backend::BackendError;
use query;
use task::{Note, Task};

/// List should be implemented by any collection of Tasks
pub trait List<'a> {
    /// Return a new List which has all completed task if yes_or_no is true and all
    /// uncompleted tasks if yes_or_no is false.
    fn completed(&mut self, yes_or_no: bool) -> Vec<Task>;
    /// Return a new List with only tasks in the given context
    fn with_context(&mut self, context: &str) -> Vec<Task>;
    /// Evaluate the AST and return a List of matching results
    fn search(&mut self, ast: query::ast::AST) -> Vec<Task>;
    /// Add a task to the List
    fn add(&mut self, task: Task) -> Result<(), BackendError>;
    /// Add multiple tasks to the List, should be more efficient resource
    /// utilization.
    fn add_multiple(&mut self, task: &mut Vec<Task>) -> Result<(), BackendError>;
    /// Return a vector of Tasks in this List
    fn into_vec(&mut self) -> Vec<Task>;
    /// Find a task by ID
    fn find_by_id(&mut self, id: &str) -> Option<&mut Task>;
    /// Return the current task, meaning the oldest uncompleted task in the List
    fn current(&mut self) -> Option<&mut Task>;

    /// Complete a task by id
    fn complete(&mut self, id: &str) -> Result<(), BackendError>;
    /// Update a task in the list, finding the original by the ID of the given task
    fn update(&mut self, task: Task) -> Result<(), BackendError>;
    /// Add note to a task by ID
    fn add_note(&mut self, id: &str, note: Note) -> Result<(), BackendError>;
}

impl<'a> List<'a> for Vec<Task> {
    fn completed(&mut self, yes_or_no: bool) -> Vec<Task> {
        self.iter()
            .filter(|t| !t.completed_date.is_none() && yes_or_no)
            .cloned()
            .collect()
    }

    fn with_context(&mut self, context: &str) -> Vec<Task> {
        self.iter()
            .filter(|t| t.context.as_str() == context)
            .cloned()
            .collect()
    }

    // TODO: implement
    fn search(&mut self, _ast: query::ast::AST) -> Self {
        self.clone()
    }

    fn add(&mut self, task: Task) -> Result<(), BackendError> {
        self.push(task);
        self.sort();
        Ok(())
    }

    fn add_multiple(&mut self, tasks: &mut Vec<Task>) -> Result<(), BackendError> {
        self.append(tasks);
        self.sort();
        Ok(())
    }

    fn into_vec(&mut self) -> Vec<Task> {
        self.clone()
    }

    fn find_by_id(&mut self, id: &str) -> Option<&mut Task> {
        for task in self {
            if task.id.as_str() == id {
                return Some(task);
            }
        }

        None
    }

    fn current(&mut self) -> Option<&mut Task> {
        self.sort();

        for task in self {
            if task.completed_date.is_none() {
                return Some(task);
            }
        }

        None
    }

    fn complete(&mut self, id: &str) -> Result<(), BackendError> {
        for task in self {
            if task.id.as_str() == id {
                match task.completed_date {
                    Some(_) => return Err(BackendError::from("Already completed.")),
                    None => {
                        task.completed_date = Some(Local::now());
                        return Ok(());
                    }
                }
            }
        }

        Err(BackendError::NotFound)
    }

    fn update(&mut self, new_task: Task) -> Result<(), BackendError> {
        for task in self.iter_mut() {
            if task.id == new_task.id {
                task.update(new_task.clone());
                return Ok(());
            }
        }

        Err(BackendError::NotFound)
    }

    fn add_note(&mut self, id: &str, note: Note) -> Result<(), BackendError> {
        for mut t in self.iter_mut() {
            if t.id.as_str() == id {
                t.notes.push(note.clone());
                return Ok(());
            }
        }

        Err(BackendError::NotFound)
    }
}

#[cfg(test)]
pub mod tests {
    use super::*;
    use task::Task;

    #[test]
    fn test_completed() {
        let mut task2 = Task::new("test2");
        task2.complete();

        let mut tasks = vec![Task::new("test1"), task2, Task::new("test3")];

        let completed = tasks.completed(true);

        assert_eq!(completed.len(), 1);
        assert_eq!(completed[0], tasks[1])
    }

    #[test]
    fn test_with_context() {
        let mut tasks = vec![
            Task::new("test1"),
            Task::new("test2").with_context("testing"),
            Task::new("test3"),
        ];

        let testing = tasks.with_context("testing");

        assert_eq!(testing.len(), 1);
        assert_eq!(testing[0], tasks[1])
    }

    //     #[test]
    //     fn test_search() {}

    #[test]
    fn test_current() {
        let mut task1 = Task::new("test1");
        task1.complete();

        let mut tasks = vec![
            task1,
            Task::new("test2"),
            Task::new("test3").with_priority(2.0),
        ];

        let mut task2 = tasks[2].clone();
        assert_eq!(tasks.current().unwrap(), &mut task2)
    }

    #[test]
    fn test_find_by_id() {
        let mut tasks = vec![Task::new("test1"), Task::new("test2"), Task::new("test3")];
        let id = tasks[2].id.clone();
        let task2 = tasks[2].clone();
        assert_eq!(task2, *tasks.find_by_id(&id).unwrap())
    }
}

File: list/list.go
// Copyright 2018 Mathew Robinson <chasinglogic@gmail.com>. All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package list

import (
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/chasinglogic/taskforge/ql/ast"
	"github.com/chasinglogic/taskforge/ql/token"
	"github.com/chasinglogic/taskforge/task"
)

// ErrNotFound is returned by a List when a task.Task matching the given ID does not
// exist
var ErrNotFound = errors.New("no task with that id exists")

// List is implemented by any struct that can maintain tasks
type List interface {
	// Init performs any initialization required for a List that does IO
	// such as creating files or connecting to databases.
	Init() error
	// Evaluate the AST and return a List of matching results
	Search(ast ast.AST) ([]task.Task, error)
	// Add a task to the List
	Add(task.Task) error
	// Add multiple tasks to the List, should be more efficient resource
	// utilization.
	AddMultiple(task []task.Task) error
	// Return a slice of task.Tasks in this List
	Slice() ([]task.Task, error)
	// Find a task by ID
	FindByID(id string) (task.Task, error)
	// Return the current task, meaning the oldest uncompleted task in the List
	Current() (task.Task, error)

	// Complete a task by id
	Complete(id string) error
	// Update a task in the listist, finding the original by the ID of the given task
	Update(task.Task) error
	// Add note to a task by ID
	AddNote(id string, note task.Note) error
}

// Config is a convenience type used to represent the unstructured config that
// will be stored by the application. All lists should accept being decoded
// using mapstructure from this type. Additionally, the Init() function should
// return an error if a required argument was not provided.
type Config map[string]interface{}

// GetByName returns the appropriate List implementation by
// a human redable string name. It returns a user friendly error
// string if not found.
func GetByName(name string) (List, error) {
	switch name {
	case "file":
		return &File{}, nil
	case "mongodb":
		fallthrough
	case "mongo":
		return &MongoList{}, nil
	default:
		return nil, fmt.Errorf("no l found with name %s", name)
	}
}

// MemoryList implements List for a slice of Tasks
type MemoryList []task.Task

// TODO: Would counting matching tasks first be faster since we
// frontload allocation?
func (ml *MemoryList) findWhere(f func(t task.Task) bool) []task.Task {
	new := []task.Task{}

	for _, t := range *ml {
		if f(t) {
			new = append(new, t)
		}
	}

	return new
}

func (ml *MemoryList) sort() {
	l := *ml

	sort.Slice(l, func(i, j int) bool {
		return (l[i].Priority > l[j].Priority) ||
			(l[i].Priority == l[j].Priority && l[i].CreatedDate.Before(l[j].CreatedDate))
	})

	ml = &l
}

// Add adds a task to this l
func (ml *MemoryList) Add(task task.Task) error {
	*ml = append(*ml, task)
	return nil
}

// AddMultiple adds multiple tasks to this l
func (ml *MemoryList) AddMultiple(tasks []task.Task) error {
	*ml = append(*ml, tasks...)
	return nil
}

// Slice turns the l into a slice of tasks
func (ml *MemoryList) Slice() ([]task.Task, error) {
	return []task.Task(*ml), nil
}

// FindByID will return a pointer to the task indicated by ID, returns
// ErrNotFound if no task with that ID exists
func (ml *MemoryList) FindByID(id string) (task.Task, error) {
	for _, t := range *ml {
		if t.ID == id {
			return t, nil
		}
	}

	return task.Task{}, ErrNotFound
}

// Current will return the first task which is not completed
func (ml *MemoryList) Current() (task.Task, error) {
	ml.sort()

	for _, t := range *ml {
		if t.CompletedDate.IsZero() {
			return t, nil
		}
	}

	return task.Task{}, ErrNotFound
}

// Complete will complete the task indicated by id
func (ml *MemoryList) Complete(id string) error {
	t, err := ml.FindByID(id)
	if err != nil {
		return err
	}

	t.CompletedDate = time.Now()
	return ml.Update(t)
}

// Update will update the task indicated by the ID of the provided task.Task
func (ml *MemoryList) Update(other task.Task) error {
	l := *ml

	for i := range l {
		if l[i].ID == other.ID {
			l[i].Context = other.Context
			l[i].Priority = other.Priority
			l[i].Body = other.Body
			l[i].Title = other.Title
			l[i].CompletedDate = other.CompletedDate

			if len(other.Notes) > len(l[i].Notes) {
				l[i].Notes = other.Notes
			}

			ml = &l
			return nil
		}
	}

	return ErrNotFound
}

// AddNote will add the note to the task indicated by the ID
func (ml *MemoryList) AddNote(id string, note task.Note) error {
	t, err := ml.FindByID(id)
	if err != nil {
		return err
	}

	t.Notes = append(t.Notes, note)
	return ml.Update(t)
}

// Search will construct a function to find matching tasks based on the given
// ast.AST
func (ml *MemoryList) Search(tree ast.AST) ([]task.Task, error) {
	return ml.findWhere(eval(tree.Expression)), nil
}

func eval(exp ast.Expression) func(t task.Task) bool {
	switch exp.(type) {
	case ast.InfixExpression:
		return evalInfixExp(exp.(ast.InfixExpression))
	case ast.StringLiteral:
		return evalStrLiteralQ(exp.(ast.StringLiteral))
	default:
		return func(t task.Task) bool { return false }
	}
}

func evalInfixExp(exp ast.InfixExpression) func(t task.Task) bool {
	switch exp.Operator.Type {
	case token.AND:
		return func(t task.Task) bool {
			return eval(exp.Left)(t) && eval(exp.Right)(t)
		}
	case token.OR:
		return func(t task.Task) bool {
			return eval(exp.Left)(t) || eval(exp.Right)(t)
		}
	case token.LIKE:
		return like(exp.Left, exp.Right)
	case token.NLIKE:
		return func(t task.Task) bool { return !like(exp.Left, exp.Right)(t) }
	case token.EQ:
		return eq(exp.Left, exp.Right)
	case token.NE:
		return func(t task.Task) bool { return !eq(exp.Left, exp.Right)(t) }
	case token.GT:
		return gt(exp.Left, exp.Right)
	case token.GTE:
		return gte(exp.Left, exp.Right)
	case token.LT:
		return func(t task.Task) bool { return !gt(exp.Left, exp.Right)(t) }
	case token.LTE:
		return lte(exp.Left, exp.Right)
	default:
		return func(t task.Task) bool { return false }
	}
}

func evalStrLiteralQ(exp ast.StringLiteral) func(t task.Task) bool {
	return func(t task.Task) bool {
		return strings.Contains(t.Title, exp.Value) ||
			strings.Contains(t.Body, exp.Value)
	}
}

func gt(left, right ast.Expression) func(task.Task) bool {
	return func(t task.Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "priority":
			return t.Priority > right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate.After(right.(ast.DateLiteral).Value)
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate.After(right.(ast.DateLiteral).Value)
		default:
			return false
		}
	}
}

func gte(left, right ast.Expression) func(task.Task) bool {
	return func(t task.Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "priority":
			return t.Priority >= right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate.After(right.(ast.DateLiteral).Value) ||
				t.CreatedDate == right.(ast.DateLiteral).Value
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate.After(right.(ast.DateLiteral).Value) ||
				t.CompletedDate == right.(ast.DateLiteral).Value
		default:
			return false
		}
	}
}

func lte(left, right ast.Expression) func(task.Task) bool {
	return func(t task.Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "priority":
			return t.Priority <= right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate.Before(right.(ast.DateLiteral).Value) ||
				t.CreatedDate == right.(ast.DateLiteral).Value
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate.Before(right.(ast.DateLiteral).Value) ||
				t.CompletedDate == right.(ast.DateLiteral).Value
		default:
			return false
		}
	}
}

func eq(left, right ast.Expression) func(task.Task) bool {
	return func(t task.Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "title":
			return t.Title == right.(ast.StringLiteral).Value
		case "body":
			return t.Body == right.(ast.StringLiteral).Value
		case "context":
			return t.Context == right.(ast.StringLiteral).Value
		case "priority":
			return t.Priority == right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate == right.(ast.DateLiteral).Value
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate == right.(ast.DateLiteral).Value
		case "completed":
			completed := right.(ast.BooleanLiteral).Value
			return t.IsCompleted() == completed
		default:
			return false
		}
	}
}

func like(left, right ast.Expression) func(task.Task) bool {
	return func(t task.Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "title":
			return strings.Contains(t.Title, right.(ast.StringLiteral).Value)
		case "body":
			return strings.Contains(t.Body, right.(ast.StringLiteral).Value)
		case "context":
			return strings.Contains(t.Context, right.(ast.StringLiteral).Value)
		default:
			return false
		}
	}
}
